File: golang/shortest_distance_to_a_character.go
package golang

func shortestToChar(s string, c byte) []int {
	var ans, indices []int
	n := len(s)

	for i := 0; i < n; i++ {
		if s[i] == c {
			indices = append(indices, i)
		}
	}

	start, end := -1, indices[0]
	indices = indices[1:]

	for i := 0; i < n; i++ {
		if start >= 0 && end >= 0 {
			ans = append(ans, Min(i-start, end-i))
		} else if start == -1 {
			ans = append(ans, end-i)
		} else {
			ans = append(ans, i-start)
		}

		if i == end {
			start = end
			if len(indices) == 0 {
				end = -1
			} else {
				end = indices[0]
				indices = indices[1:]
			}

		}
	}

	return ans
}

File: python3/shortest_distance_to_a_character.py
"""
Space   : O(n)
Time    : O(n)
"""


class Solution:
    def shortestToChar(self, s: str, c: str) -> List[int]:
        ans, indices = [], []
        n = len(s)

        for i in range(n):
            if s[i] == c:
                indices.append(i)

        start, end = -1, indices.pop(0)

        for i in range(n):
            if start >= 0 and end >= 0:
                ans.append(min(i - start, end - i))
            elif start == -1:
                ans.append(end - i)
            elif end == -1:
                ans.append(i - start)

            if i == end:
                start = end
                if not indices:
                    end = -1
                else:
                    end = indices.pop(0)

        return ans
